staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+				"mapOfMap": mapType(mapTypePtr(&stringType)),
+				"mapOfObj": mapType(objectTypePtr(map[string]schema.Structural{
+					"field2": stringType,
+				})),
+				"mapOfListMap": mapType(listMapTypePtr([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k": stringType,
+					"v": stringType,
+				}))),
+				"mapOfList": mapType(listTypePtr(&stringType)),
+				"listMapOfObj": listMapType([]string{"k2"}, objectTypePtr(map[string]schema.Structural{
+					"k2": stringType,
+					"v2": stringType,
+				})),
+				"listOfMap": listType(mapTypePtr(&stringType)),
+				"listOfObj": listType(objectTypePtr(map[string]schema.Structural{
+					"field3": stringType,
+				})),
+				"listOfListMap": listType(listMapTypePtr([]string{"k3"}, objectTypePtr(map[string]schema.Structural{
+					"k3": stringType,
+					"v3": stringType,
+				}))),
+			}),
+			valid: []string{
+				"obj.field == 'a'",
+				"mapOfMap['x']['y'] == 'b'",
+				"mapOfObj['k'].field2 == 'c'",
+				"mapOfListMap['o'].exists(e, e.k == '1' && e.v == 'd')",
+				"mapOfList['l'][0] == 'e'",
+				"listMapOfObj.exists(e, e.k2 == '2' && e.v2 == 'f')",
+				"listOfMap[0]['z'] == 'g'",
+				"listOfObj[0].field3 == 'h'",
+				"listOfListMap[0].exists(e, e.k3 == '3' && e.v3 == 'i')",
+			},
+			errors: map[string]string{},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			for _, validRule := range tt.valid {
+				s := withRule(*tt.schema, validRule)
+				celValidator := NewValidator(&s)
+				errs := celValidator.Validate(field.NewPath("root"), &s, tt.obj)
+				for _, err := range errs {
+					t.Errorf("unexpected error: %v", err)
+				}
+			}
+			for rule, expectErrToContain := range tt.errors {
+				s := withRule(*tt.schema, rule)
+				celValidator := NewValidator(&s)
+				errs := celValidator.Validate(field.NewPath("root"), &s, tt.obj)
+				if len(errs) == 0 {
+					t.Error("expected validation errors but got none")
+				}
+				for _, err := range errs {
+					if err.Type != field.ErrorTypeInvalid || !strings.Contains(err.Error(), expectErrToContain) {
+						t.Errorf("expected error to contain '%s', but got: %v", expectErrToContain, err)
+					}
+				}
+			}
+
+		})
+	}
+}
+
+func primitiveType(typ, format string) schema.Structural {
+	result := schema.Structural{
+		Generic: schema.Generic{
+			Type: typ,
+		},
+	}
+	if len(format) != 0 {
+		result.ValueValidation = &schema.ValueValidation{
+			Format: format,
+		}
+	}
+	return result
+}
+
+var (
+	integerType = primitiveType("integer", "")
+	int32Type   = primitiveType("integer", "int32")
+	int64Type   = primitiveType("integer", "int64")
+	numberType  = primitiveType("number", "")
+	floatType   = primitiveType("number", "float")
+	doubleType  = primitiveType("number", "double")
+	stringType  = primitiveType("string", "")
+	byteType    = primitiveType("string", "byte")
+	binaryType  = primitiveType("string", "binary")
+	booleanType = primitiveType("boolean", "")
+	dateType    = primitiveType("date", "")
```

Comment:
`primitiveType("string", "date)`

Reply:
I do have the "date" string format above. This date type was added accidentally. Removing.